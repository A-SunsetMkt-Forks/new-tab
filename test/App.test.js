'use strict';

require('jest-canvas-mock');
const App = require('../src/App.html');

describe('App root component', () => {
  it('renders correctly', () => {
    const target = document.createElement('div');
    new App({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });
});

describe('App click handler', () => {
  it('triggers on mouse click event', () => {
    expect.assertions(1);
    const target = document.createElement('div');
    const component = new App({ target });
    const spy = jest.spyOn(component, '_onLinkClick');
    const event = new MouseEvent('click');
    window.dispatchEvent(event);
    expect(spy).toHaveBeenCalledWith(event);
    spy.mockReset();
    spy.mockRestore();
  });

  it('handles regular HTTP links', () => {
    expect.assertions(2);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="https://github.com/MaxMilton/new-tab/issues"]');
    const event = new MouseEvent('click', {
      view: window,
      bubbles: true,
      cancelable: true,
    });
    Object.defineProperty(event, 'target', { value: link, enumerable: true });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).not.toHaveBeenCalled();
    spy1.mockReset();
    spy2.mockReset();
    spy1.mockRestore();
    spy2.mockRestore();
  });

  it('handles chrome internal links', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="chrome://bookmarks/"]');
    const event = new MouseEvent('click', {
      view: window,
      bubbles: true,
      cancelable: true,
    });
    Object.defineProperty(event, 'target', { value: link, enumerable: true });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    const spy3 = jest.spyOn(chrome.tabs, 'create', 'get');
    const spy4 = jest.spyOn(chrome.tabs, 'update', 'get');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).not.toHaveBeenCalled();
    expect(spy4).toHaveBeenCalled();
    spy1.mockReset();
    spy2.mockReset();
    spy3.mockReset();
    spy4.mockReset();
    spy1.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
    spy4.mockRestore();
  });

  it('handles chrome internal links with _blank target', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="chrome://bookmarks/"]');
    const event = new MouseEvent('click', {
      view: window,
      bubbles: true,
      cancelable: true,
    });
    Object.defineProperty(event, 'target', { value: link, enumerable: true });
    Object.defineProperty(event.target, 'target', { value: '_blank', enumerable: true });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    const spy3 = jest.spyOn(chrome.tabs, 'create', 'get');
    const spy4 = jest.spyOn(chrome.tabs, 'update', 'get');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).toHaveBeenCalled();
    expect(spy4).not.toHaveBeenCalled();
    spy1.mockReset();
    spy2.mockReset();
    spy3.mockReset();
    spy4.mockReset();
    spy1.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
    spy4.mockRestore();
  });

  it('handles chrome internal links with ctrl key', () => {
    expect.assertions(4);
    const target = document.createElement('div');
    const component = new App({ target });
    const link = target.querySelector('[href="chrome://bookmarks/"]');
    const event = new MouseEvent('click', {
      view: window,
      bubbles: true,
      cancelable: true,
      ctrlKey: true,
    });
    Object.defineProperty(event, 'target', { value: link, enumerable: true });
    const spy1 = jest.spyOn(component, '_onLinkClick');
    const spy2 = jest.spyOn(event, 'preventDefault');
    const spy3 = jest.spyOn(chrome.tabs, 'create', 'get');
    const spy4 = jest.spyOn(chrome.tabs, 'update', 'get');
    window.dispatchEvent(event);
    expect(spy1).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    expect(spy3).toHaveBeenCalled();
    expect(spy4).not.toHaveBeenCalled();
    spy1.mockReset();
    spy2.mockReset();
    spy3.mockReset();
    spy4.mockReset();
    spy1.mockRestore();
    spy2.mockRestore();
    spy3.mockRestore();
    spy4.mockRestore();
  });
});
