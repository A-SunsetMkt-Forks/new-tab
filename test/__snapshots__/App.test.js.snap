// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App root component renders correctly 1`] = `
"<div id=\\"a\\"><div id=\\"b\\"></div>

  <a href=\\"undefined\\" title=\\"Other bookmarks\\" class=\\"b\\">
    <span class=\\"t\\">Other bookmarks</span></a><!----></div>
<div id=\\"m\\"><div id=\\"i\\">â˜°</div>

  <div id=\\"d\\"><a href=\\"chrome-search://local-ntp/local-ntp.html\\" class=\\"m\\">Open Default Tab</a>
    <a href=\\"chrome://bookmarks/\\" class=\\"m\\">Bookmarks Manager</a>
    <a href=\\"chrome://downloads/\\" class=\\"m\\">Downloads</a>
    <a href=\\"chrome://history/\\" class=\\"m\\">History</a>
    <a href=\\"chrome://settings/passwords\\" class=\\"m\\">Passwords</a>

    <hr>

    <a href=\\"#\\" class=\\"m\\">New Tab Settings</a>
    <a href=\\"https://github.com/MaxMilton/new-tab/issues\\" class=\\"m\\">Submit Bug</a></div></div>
<div class=\\"c\\"><input id=\\"s\\" placeholder=\\"Search browser...\\" autocomplete=\\"off\\">

  
  <p class=\\"g\\">No matching Open Tabs</p>

  
  

  
  

  
  </div>"
`;
